WIP [Design] v1beta1 API (#2997)
jetstack/cert-manager: design/20200610.v1beta1api.md

---
title: v1beta1 API
authors:
  - "@meyskens"
reviewers:
  - "@joshvanl"
  - "@munnerz"
approvers:
  - "@joshvanl"
  - "@munnerz"
creation-date: 2020-06-10
last-updated: 2020-06-10
---

# v1beta1 API

<!-- toc -->
- [Summary](#summary)
- [Motivation](#motivation)
  - [Goals](#goals)
  - [Non-Goals](#non-goals)
- [Proposal](#proposal)
  - [API Changes](#api-changes)
  - [Legacy version](#legacy-version)
<!-- /toc -->

## Summary

Moving towards cert-manager v1 we have to mature our API. Currently being on v1aplha3 the next step would be v1beta1.

## Motivation

> With cert-manager maturing and seeing bugger user adoption we're working towards a stable release. This involves having a stable API version.

Reviewer: ```suggestion
With cert-manager maturing and seeing bigger user adoption we're working towards a stable release. This involves having a stable API version.
```

We saw some big changes between v1alpha1 and v1alpha2, with v1alpha3 as a small improvement.
v1alpha3 is generally fitting our needs, however it has a few shortcomings we should address when maturing the API.
With the introduction of v1alpha3 also came the conversion webhooks into cert-manager. These enable us to give our users a frictionless
path to upgrade their cert-manager resources. 

### Goals

* Improve the ACME API to support fixing challenges in the past
* Improve our API with small breaking changes
* Provide a frictionless upgrade path

### Non-goals

* We must not make a total redesign of the API

### API changes

#### acme.cert-manager.io
ACME Issuers make use of several ACME specific cert-manager resources to store it's state. We have challenges.acme.cert-manager.io and orders.acme.cert-manager.io.

> An ACME order is created to call the ACME instance to request a certificate. This created a challenge resource which than is used by cert-manager to prove the domain's ownership to the ACME issuer.

Reviewer: ```suggestion
An ACME order is created to call the ACME instance to request a certificate. This created a challenge resource which then is used by cert-manager to prove the domain's ownership to the ACME issuer.
```

> In [RFC 8555](#TODO) it is specified that one ACME order results in one or more Authorizations needed to be fufulled, one authorization also can have one or more challenges.

Reviewer: ```suggestion
In [RFC 8555](https://tools.ietf.org/html/rfc8555) it is specified that one ACME order results in one or more Authorizations needed to be fulfilled, one authorization also can have one or more challenges.
```

> cert-manager currently abstracts the authorization resource away. In [an issue](https://github.com/jetstack/cert-manager/issues/2494#issuecomment-622340474) it became apperent 

Reviewer: ```suggestion
cert-manager currently abstracts the authorization resource away. In [an issue](https://github.com/jetstack/cert-manager/issues/2494#issuecomment-622340474) it became apparent 
```

> that there is a case where an authotization can already be valid and thus needs no challenges to be created, this makes some weird side effects when not storing authotizations.

Reviewer: ```suggestion
that there is a case where an authorization can already be valid and thus needs no challenges to be created, this makes some weird side effects when not storing authorizations.
```

##### authotization.acme.cert-manager.io

> We should add an authotization resource just like order to cert-manager.

Reviewer: ```suggestion
We should add an authorization resource just like order to cert-manager.
```

```yaml
apiVersion: acme.cert-manager.io/v1beta1

> kind: Authotization

Reviewer: ```suggestion
kind: Authorization
```
